solver/llbsolver/vertex.go

```diff
@@ -188,34 +189,90 @@ func newVertex(dgst digest.Digest, op *pb.Op, opMeta *pb.OpMetadata, load func(d
 
 // loadLLB loads LLB.
 // fn is executed sequentially.
-func loadLLB(def *pb.Definition, fn func(digest.Digest, *pb.Op, func(digest.Digest) (solver.Vertex, error)) (solver.Vertex, error)) (solver.Edge, error) {
+func loadLLB(ctx context.Context, def *pb.Definition, polEngine SourcePolicyEvaluator, fn func(digest.Digest, *pb.Op, func(digest.Digest) (solver.Vertex, error)) (solver.Vertex, error)) (solver.Edge, error) {
 	if len(def.Def) == 0 {
 		return solver.Edge{}, errors.New("invalid empty definition")
 	}
 
 	allOps := make(map[digest.Digest]*pb.Op)
+	mutatedDigests := make(map[digest.Digest]digest.Digest) // key: old, val: new
 
-	var dgst digest.Digest
+	var lastDgst digest.Digest
 
 	for _, dt := range def.Def {
 		var op pb.Op
 		if err := (&op).Unmarshal(dt); err != nil {
 			return solver.Edge{}, errors.Wrap(err, "failed to parse llb proto op")
 		}
-		dgst = digest.FromBytes(dt)
+		dgst := digest.FromBytes(dt)
+		if polEngine != nil {
+			mutated, err := polEngine.Evaluate(ctx, &op)
+			if err != nil {
+				return solver.Edge{}, errors.Wrap(err, "error evaluating the source policy")
+			}
+			if mutated {
+				dtMutated, err := op.Marshal()
+				if err != nil {
+					return solver.Edge{}, err
+				}
+				dgstMutated := digest.FromBytes(dtMutated)
+				mutatedDigests[dgst] = dgstMutated
+				dgst = dgstMutated
+			}
+		}
 		allOps[dgst] = &op
+		lastDgst = dgst
+	}
+
+	recompute := func(mutatedDigests map[digest.Digest]digest.Digest) (map[digest.Digest]digest.Digest, error) {
+		// Rewrite the input digests if some op was mutated
+		newMut := make(map[digest.Digest]digest.Digest)
+		for dgst, op := range allOps {
+			var mutated bool
+			for _, inp := range op.Inputs {
+				if mutatedDgst, ok := mutatedDigests[inp.Digest]; ok {
+					inp.Digest = mutatedDgst
+					mutated = true
+				}
+			}
+			if mutated {
+				dt, err := op.Marshal()
+				if err != nil {
+					return nil, err
+				}
+				newDgst := digest.FromBytes(dt)
+				allOps[newDgst] = op
+				delete(allOps, dgst)
+				newMut[dgst] = newDgst
+				if dgst == lastDgst {
+					lastDgst = newDgst
+				}
+			}
+		}
+		return newMut, nil
+	}
+
+	for {
+		var err error
+		mutatedDigests, err = recompute(mutatedDigests)
```

Comment:
This seems to have an exponential complexity. Could we instead do it with a recursive function with a `visited map[digest]digest` that makes sure that no vertex is checked twice?

Reply:
Updated.